http/cookies.lua

```diff
@@ -0,0 +1,353 @@
+		cookie[#cookie + 1] = "; Expires=" .. util.imf_date(data.expires)
+	end
+	if data.max_age then
+		cookie[#cookie + 1] = "; Max-Age=" .. string.format("%d", data.max_age)
+	end
+	if data.domain then
+		cookie[#cookie + 1] = "; Domain=" .. data.domain
+	end
+	if data.path then
+		cookie[#cookie + 1] = "; Path=" .. util.encodeURI(data.path)
+	end
+	if data.secure then
+		cookie[#cookie + 1] = "; Secure"
+	end
+	if data.http_only then
+		cookie[#cookie + 1] = "; HttpOnly"
+	end
+	-- This component is not a part of the RFC 6265 specification for the
+	-- headers, but is instead from a draft of another RFC that builds on the
+	-- original one.
+	-- https://tools.ietf.org/html/draft-ietf-httpbis-cookie-same-site-00#section-4.1
+	if data.same_site then
+		local v
+		if data.same_site:lower() == "strict" then
+			v = "; SameSite=Strict"
+		elseif data.same_site:lower() == "lax" then
+			v = "; SameSite=Lax"
+		else
+			error('invalid value for same_site, expected "Strict" or "Lax"')
+		end
+		cookie[#cookie + 1] = v
+	end
+	return table.concat(cookie)
+end
+
+local Cookie_anchored = http_patts.Cookie * lpeg.P(-1)
+
+local function match_cookies(cookie)
+	local match = Cookie_anchored:match(cookie)
+	if match then
+		return match
+	else
+		return nil, "improper Cookie header format"
+	end
+end
+
+local function parse_cookies(cookie)
+	local cookies = match_cookies(cookie)
+	local to_add = {}
+	for k, v in pairs(cookies) do
+		to_add[#to_add + 1] = {k, v}
+	end
+	local len = #cookies
+	for i, v in ipairs(to_add) do
+		cookies[len + i] = v
+	end
+	table.sort(cookies, function(t1, t2)
+		return t1[1] < t2[1]
+	end)
+	return cookies
+end
+
+local cookiejar_methods = {}
+if psl.latest then
+	cookiejar_methods.psl_object = psl.latest()
+else
+	-- older versions of libpsl do not offer a `latest` list
+	cookiejar_methods.psl_object = psl.builtin()
+end
+local cookiejar_mt = {
+	__name = "http.cookies.cookiejar";
+	__index = cookiejar_methods;
+}
+
+local function new_cookiejar()
+	return setmetatable({cookies={}}, cookiejar_mt)
+end
+
+function cookiejar_methods:add(cookie, time)
+	cookie.last_access = time or os.time()
+	local domain, path, key = cookie.domain, cookie.path, cookie.key
+	local cookies = self.cookies
+	if cookies[domain] and cookies[domain][path] then
+		local old_cookie = cookies[domain][path][key]
+		if old_cookie then
+			cookie.creation = old_cookie.creation
+		end
+	end
+
+	local old_cookie = self:get(cookie.domain, cookie.path, cookie.key)
+	if old_cookie then
+		self:remove_cookie(old_cookie)
+	end
+	if cookie.persistent then
+		local cookie_exp_time = cookie.expires
+		local inserted = false
+		for i=1, #cookies do
+			-- insert into first spot where cookie expires after
+			if cookies[i].expires < cookie_exp_time then
+				inserted = true
+				table.insert(cookies, i, cookie)
+			end
+		end
+		if not inserted then
+			cookies[#cookies + 1] = cookie
+		end
+	else
+		cookie.expires = math.huge
+		table.insert(cookies, 1, cookie)
```

Comment:
*TODO:* Add explanation; the "older" a nonpermanent cookie is, the more likely it should be nuked, hence why it's put in position 1